src/treelearner/cost_effective_gradient_boosting.hpp

```diff
@@ -63,6 +64,17 @@ class CostEfficientGradientBoosting {
     }
     init_ = true;
   }
+
+  void BeforeTrain() {
+    // clear the splits in splits_per_leaf_
+    const int num_total_splits = static_cast<int>(splits_per_leaf_.size());
```

Comment:
I believe this can overflow on wide datasets:
https://github.com/microsoft/LightGBM/blob/3d25e373ac09fa0cd5ce75a9decbc195e0bd5220/src/treelearner/cost_effective_gradient_boosting.hpp#L38-L40
https://github.com/microsoft/LightGBM/blob/3d25e373ac09fa0cd5ce75a9decbc195e0bd5220/include/LightGBM/config.h#L181-L183
More precisely, on datasets with `16 384` features and maximum allowed `num_leaves` (`131072`) value.
I believe 16k features are not something unreachable in a user scenarios and some autoML/hyperopt software may use maximum allowed value for any parameter.

Reply:
Thanks for spotting this! The size is cast into signed int because omp does not support unsigned int as index in for loop. 
I've switched to `Threading::For` to avoid this.

Reply:
Resolved via 45de0a4

Reply:
Thanks a lot for doing this!